extensions/wikia/Insights/specials/PagesWithoutInfobox.class.php

```diff
@@ -0,0 +1,140 @@
+		parent::__construct( $name );
+	}
+
+	public function isListed() {
+		return false;
+	}
+
+	public function isExpensive() {
+		return true;
+	}
+
+	/**
+	 * A wrapper for calling the querycache table
+	 *
+	 * @param bool $offset
+	 * @param int $limit
+	 * @return ResultWrapper
+	 */
+	public function doQuery( $offset = false, $limit = self::LIMIT ) {
+		return $this->fetchFromCache( $limit, $offset );
+	}
+
+	/**
+	 * Update the querycache table
+	 *
+	 * @param bool $limit Only for consistency
+	 * @param bool $ignoreErrors Only for consistency
+	 * @return bool|int
+	 */
+	public function recache( $limit = false, $ignoreErrors = true ) {
+		$dbw = wfGetDB( DB_MASTER );
+
+		/**
+		 * 1. Get the new data first
+		 */
+		$pagesWithoutInfobox = $this->reallyDoQuery();
+		$dbw->begin();
+
+		/**
+		 * 2. Delete the existing records
+		 */
+		( new WikiaSQL() )
+			->DELETE( 'querycache' )
+			->WHERE( 'qc_type' )->EQUAL_TO( $this->getName() )
+			->run( $dbw );
+
+		/**
+		 * 3. Insert the new records if the $pagesWithoutInfobox array is not empty
+		 */
+		$num = 0;
+		if ( !empty( $pagesWithoutInfobox ) ) {
+
+			( new WikiaSQL() )
+				->INSERT()->INTO( 'querycache', [
+					'qc_type',
+					'qc_value',
+					'qc_namespace',
+					'qc_title'
+				] )
+				->VALUES( $pagesWithoutInfobox )
+				->run( $dbw );
+
+			$num = $dbw->affectedRows();
+			if ( $num === 0 ) {
+				$dbw->rollback();
+				$num = false;
+			} else {
+				$dbw->commit();
+			}
+		}
+
+		wfRunHooks( 'PagesWithoutInfoboxQueryRecached' );
+
+		return $num;
+	}
+
+	/**
+	 * Returns an array with data on pages that do not have an infobox.
+	 *
+	 * @param bool $limit Only for consistency
+	 * @param bool $offset Only for consistency
+	 * @return bool|mixed
+	 */
+	public function reallyDoQuery( $limit = false, $offset = false ) {
+		global $wgCityId, $wgContentNamespaces;
+
+		$tc = new TemplateClassificationService();
+
+		$infoboxTemplates = [];
+		foreach( $tc->getTemplatesOnWiki( $wgCityId ) as $pageId => $templateType ) {
+			if ( $templateType === TemplateClassificationService::TEMPLATE_INFOBOX ) {
+				$templateTitle = Title::newFromID( $pageId );
+				if ( $templateTitle instanceof Title ) {
+					$infoboxTemplates[] = $templateTitle->getDBkey();
+				}
+			}
+		}
+
+		$dbr = wfGetDB( DB_SLAVE, [ $this->getName(), __METHOD__, 'vslow' ] );
+
+		$pagesWithInfobox = [];
+		if ( !empty( $infoboxTemplates ) ) {
+			$pagesWithInfobox = ( new WikiaSQL() )
+				->SELECT( 'tl_from' )
+				->FROM( 'templatelinks' )
+				->WHERE( 'tl_title' )->IN( $infoboxTemplates )
+				->runLoop( $dbr, function ( &$pagesWithInfobox, $row ) {
+					$pagesWithInfobox[] = $row->tl_from;
+				} );
+		}
+
+		$sql = ( new WikiaSQL() )
+			->SELECT( 'page_id', 'page_title', 'page_namespace' )
+			->FROM( 'page' )
+			->WHERE( 'page_namespace' )->IN( $wgContentNamespaces );
+
+		if ( !empty( $pagesWithInfobox ) ) {
+			$sql->AND_( 'page_id' )->NOT_IN( $pagesWithInfobox );
```

Comment:
This query can have an arbitrary number of pages in it's NOT_IN clause -- may not be wise.

@adamkarminski please do an explain here and ping @Grunny and/or @wladekb